contributors/design-proposals/service_links.md

```diff
@@ -0,0 +1,237 @@
+# Explicit service links
+
+## Abstract
+
+Proposes an API for explicit links to services in the pod spec.
+
+## Motivation
+
+Currently, all pods are automatically injected with docker-link-style
+environment variables for each service in their namespaces.  Making service
+dependencies explicit in the PodSpec has a number of advantages:
+
+1.  Reduce accidental coupling
+2.  Reduce the number of IP tables rules that have to be created by the
+    kube-proxy
+3.  Identify relationships between pods and services, allowing user interfaces
+    to display additional information for end users.
```

Comment:
Just observations for now. Will comment with recommendations later.

We have a few things potentially going on here:
* Defer starting pods/containers until consumed services exist
* Don't create iptables rules unless some pod depends on them
* Only create service env vars upon request

kubectl create and apply bulk-create resources in arbitrary order. (Back at the dawn of the K8s universe, kubecfg created services first.)

iptables rules get created on every node for all services in the cluster as soon as kube-proxy observes the services. When kube-proxy is down, they don't get created/updated, but otherwise it's likely the rules will be created prior to creation of pods consuming the services, assuming the services are created before or immediately after the controllers that will generate the consuming pods, which then have to be created, scheduled, pulled, and started by kubelet.

Kubelet injects service env vars for all services in the same namespace upon container creation:
https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/envvars/envvars.go#L30

kube-dns watches Services and Endpoints, and updates its own state as it observes changes.

This all occurs asynchronously.

We could influence creation order with more general-purpose inter-resource dependencies. In the current proposal, the workload controllers and services (and ingress) would be at the same "application" layer of the system, so controllers could also defer pod creation until consumed services existed, or even ready (assuming we introduce service readiness).

Env vars could also be injected at the application layer or policy layer, but an explicit knob would need to be introduced into pods to disable the current (layer-violating) injection.

iptables modifications couldn't be changed without a field or annotation on the pods. Note also that doing so would increase the likelihood of rules creation lagging container start without a handshake with kubelet to prevent that.

Additionally, in the layers doc, we discussed that serviceAccount could be pushed to a higher layer than core despite the pod reference.